uamqp/session.py

```diff
@@ -214,13 +215,61 @@ def _outgoing_transfer(self, delivery):
         if self.remote_incoming_window <= 0:
             delivery.transfer_state = SessionTransferState.Busy
         else:
+            payload = delivery.frame['payload']
+            payload_size = len(payload)
+
             delivery.frame['delivery_id'] = self.next_outgoing_id
-            self._connection._process_outgoing_frame(self.channel, TransferFrame(**delivery.frame))
+            # calculate the transfer frame encoding size excluding the payload
+            delivery.frame['payload'] = b""
+            encoded_frame = encode_frame(TransferFrame(**delivery.frame))[1]
+            transfer_overhead_size = len(encoded_frame)
+
+            # available size for payload per frame is calculated as following:
+            # remote max frame size - transfer overhead (calculated) - header (8 bytes)
+            available_frame_size = self._connection._remote_max_frame_size - transfer_overhead_size - 8
+
+            start_idx = 0
+            remaining_payload_cnt = payload_size
+            # encode n-1 frames if payload_size > available_frame_size
+            while remaining_payload_cnt > available_frame_size:
+                tmp_delivery_frame = {
+                    'handle': delivery.frame['handle'],
+                    'delivery_tag': delivery.frame['delivery_tag'],
+                    'message_format': delivery.frame['message_format'],
+                    'settled': delivery.frame['settled'],
+                    'more': True,
+                    'rcv_settle_mode': delivery.frame['rcv_settle_mode'],
+                    'state': delivery.frame['state'],
+                    'resume': delivery.frame['resume'],
+                    'aborted': delivery.frame['aborted'],
+                    'batchable': delivery.frame['batchable'],
+                    'payload': payload[start_idx:start_idx+available_frame_size],
+                    'delivery_id': self.next_outgoing_id
+                }
+                self._connection._process_outgoing_frame(self.channel, TransferFrame(**tmp_delivery_frame))
+                start_idx += available_frame_size
+                remaining_payload_cnt -= available_frame_size
+
+            # encode the last frame
+            tmp_delivery_frame = {
+                'handle': delivery.frame['handle'],
+                'delivery_tag': delivery.frame['delivery_tag'],
+                'message_format': delivery.frame['message_format'],
+                'settled': delivery.frame['settled'],
+                'more': False,
```

Comment:
What if the loop above leaves 0 available frame size? That would mean this frame is empty. If we don't send it, then `more` is never set to `False`.

Reply:
do you mean `available_frame_size` could be 0 that leads to an infinite loop of `while remaining_payload_cnt > available_frame_size:`?

I'm thinking that shouldn't this be an invalid case that we should raise error..

Reply:
No, my thinking was rather - what if payload size is exactly divisible by `available_frame_size`?
For example:
```python
start_index = 0
remaining_payload_cnt = 100
available_frame_size = 50
while remaining_payload_cnt > available_frame_size:
    remaining_payload_cnt -= available_frame_size
    start_index += available_frame_size

# Once the loop has finished we are left with:
# remaining_payload_cnt = 0
# start_index = 100

tmp_delivery_frame = {
                'payload': payload[start_idx:],  # This will be empty
                'more': False,  # Now we have to 'close' the multi-frame with more=False, even though we have no more payload.
}
```

Reply:
aha, in that case, the while loop will be executed one time only.

in the loop, remaining_payload_cnt will be 100 - 50 = 50 which doesn't satisfy the condition `while remaining_payload_cnt > available_frame_size:` (while 50 > 50)

Reply:
Hah - your right - my example is assuming `>=` ^_^;;
Should be good then!